lib/ContextParser.ts

```diff
@@ -760,46 +789,48 @@ must be one of ${Util.CONTAINERS.join(', ')}`, ERROR_CODES.INVALID_CONTAINER_MAP
       }
 
       this.applyScopedProtected(importContext, { processingMode }, defaultExpandOptions);
+
       let newContext: IJsonLdContextNormalizedRaw = { ...importContext, ...context };
+
+      // Handle terms (before protection checks)
+      this.idifyReverseTerms(newContext);
+      this.normalize(newContext, { processingMode, normalizeLanguageTags });
+      this.applyScopedProtected(newContext, { processingMode }, defaultExpandOptions);
+
+      const keys = Object.keys(newContext);
       if (typeof parentContext === 'object') {
         // Merge different parts of the final context in order
-        this.applyScopedProtected(newContext, { processingMode }, defaultExpandOptions);
         newContext = { ...parentContext, ...newContext };
       }
 
-      const newContextWrapped = new JsonLdContextNormalized(newContext);
-
       // Parse inner contexts with minimal processing
-      await this.parseInnerContexts(newContext, options);
+      await this.parseInnerContexts(newContext, options, keys);
+
+      const newContextWrapped = new JsonLdContextNormalized(newContext);
 
       // In JSON-LD 1.1, @vocab can be relative to @vocab in the parent context, or a compact IRI.
       if ((newContext && newContext['@version'] || ContextParser.DEFAULT_PROCESSING_MODE) >= 1.1
         && ((context['@vocab'] && typeof context['@vocab'] === 'string') || context['@vocab'] === '')) {
         if (parentContext && '@vocab' in parentContext && context['@vocab'].indexOf(':') < 0) {
           newContext['@vocab'] = parentContext['@vocab'] + context['@vocab'];
-        } else {
-          if (Util.isCompactIri(context['@vocab']) || context['@vocab'] in newContextWrapped.getContextRaw()) {
+        } else if (Util.isCompactIri(context['@vocab']) || context['@vocab'] in newContext) {
             // @vocab is a compact IRI or refers exactly to a prefix
-            newContext['@vocab'] = newContextWrapped.expandTerm(context['@vocab'], true);
-          }
+          newContext['@vocab'] = newContextWrapped.expandTerm(context['@vocab'], true);
+
         }
       }
 
-      // Handle terms (before protection checks)
-      this.idifyReverseTerms(newContext);
+      // FIXME: Add keys as a 3rd argument here for performance
       this.expandPrefixedTerms(newContextWrapped, this.expandContentTypeToBase);
```

Comment:
When we add the `keys` argument here (which only iterates over the new context keys added and so skips old parent context keys), we get a further ~30% performance improvement with the following result when we make this change on top of https://github.com/rubensworks/jsonld-context-parser.js/commit/84a1e4199a34caa43e392d91ed340f2c2fc4f2de.

In https://github.com/rubensworks/jsonld-context-parser.js/commit/84a1e4199a34caa43e392d91ed340f2c2fc4f2de:
 Parse a context that has not been cached; and without caching in place x **679** ops/sec ±0.78% (84 runs sampled)
 
After making this change:
Parse a context that has not been cached; and without caching in place x **857** ops/sec ±2.75% (90 runs sampled)

All of the unit tests in this library pass with this change, as do all of the spec tests in `jsonld-streaming-parser`. However the following unit test from `jsonld-streaming-parser` breaks with this change:

```bash
  ● ParsingContext › an empty instance › getContext › for type-scoped and property-scoped contexts › should consider the applicable property-scoped context

    expect(received).toEqual(expected) // deep equality

    - Expected  - 3
    + Received  + 1

    @@ -2,13 +2,11 @@
        "contextRaw": Object {
          "@__propagateFallback": Object {
            "fallback": true,
          },
          "@vocab": "http://bla.org/",
    -     "bar": Object {
    -       "@id": "http://bla.org/bar",
    -     },
    +     "bar": Object {},
          "baz": Object {
            "@id": "http://bla.org/baz",
            "@type": "@vocab",
          },
        },

      622 |         it('should consider the applicable property-scoped context', async () => {
      623 |           return expect(await parsingContext.getContext(['', 'a', 'bar'], 0))
    > 624 |             .toEqual(new JsonLdContextNormalized({
          |              ^
      625 |               '@__propagateFallback': { fallback: true },
      626 |               '@vocab': 'http://bla.org/',
      627 |               'bar': { '@id': 'http://bla.org/bar' },

      at Object.<anonymous> (test/ParsingContext-test.ts:624:14)

```

Reply:
It looks to me like the unit test is correct to expect that. Something that can be easily fixed?

Reply:
Oh, I see.
Ok, in that case we can modify the unit test indeed so that the parent context is expanded.

> I think this might only be a problem if the parent context has been produced from a .parse call with minimalProcessing; which appears to be a mode that the jsonld-streaming-parser never opts into (unless there is a way of users configuring that parameter)?

Indeed, this mode is only for validation. The streaming parser will never use it.